src/Perpetuum/Zones/Effects/ZoneEffects/ZoneEffectHandler.cs

```diff
@@ -1,22 +1,86 @@
-﻿using Perpetuum.Players;
+﻿using Perpetuum.Log;
+using Perpetuum.Players;
+using Perpetuum.Units;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
 
 namespace Perpetuum.Zones.Effects.ZoneEffects
 {
     public class ZoneEffectHandler : IZoneEffectHandler
     {
-        private readonly IZoneEffectRepository _zoneEffectRepository;
-        public ZoneEffectHandler(IZoneEffectRepository zoneEffectRepository)
+        private const byte _ = 0;
+        private readonly IZone _zone;
+        private readonly ConcurrentDictionary<ZoneEffect, byte> _effects;
+
+        public ZoneEffectHandler(IZone zone)
+        {
+            _zone = zone;
+            _effects = new ConcurrentDictionary<ZoneEffect, byte>();
+            InitCollection();
+        }
+
+        private void InitCollection()
+        {
+            foreach (var zoneEffect in ZoneEffectReader.GetStaticZoneEffects(_zone))
+            {
+                _effects.Add(zoneEffect, _);
+            }
+        }
+
+        public void AddEffect(ZoneEffect zoneEffect)
+        {
+            _effects.TryAdd(zoneEffect, _);
+            OnZoneEffectAdded(zoneEffect);
+        }
+
+        public void RemoveEffect(ZoneEffect zoneEffect)
+        {
+            _effects.TryRemove(zoneEffect, out byte b);
+            OnZoneEffectRemoved(zoneEffect);
+        }
+
+        private void OnZoneEffectRemoved(ZoneEffect effect)
+        {
+            var eligibleUnits = _zone.Units.Where(u => u.EffectHandler.ContainsEffect(effect.Effect));
+            foreach (var unit in eligibleUnits)
+            {
+                unit.EffectHandler.RemoveEffectsByType(effect.Effect);
+                Logger.DebugInfo($"Removing {effect.Effect} to {unit} on zone:{_zone.Id}");
+            }
+        }
+
+        private void OnZoneEffectAdded(ZoneEffect effect)
+        {
+            var eligibleUnits = _zone.Units.Where(u => CanApplyEffect(effect)(u));
```

Comment:
_zone.Units.Where(u => CanApplyEffect(effect)**(u)**); does this **(u)** compile? What does it do?

Reply:
Look at the method below on line 70. `CanApplyEffect` returns a function that itself returns a boolean. So what's happening here is that `u` is getting passed to the return value (function) of `CanApplyEffect` and then the result of *that* function is what is used for the `where` filter.